File: minimum-depth-of-binary-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
// 0 nodes = 0
// 1 node = 1
// and so on. 
// so bottom-up: min of left and right is 1 + the min you can get
func minDepth(root *TreeNode) int {
    if root == nil { return 0 }
    if root.Left != nil && root.Right != nil {
        return 1 + min(minDepth(root.Left), minDepth(root.Right))
    }
    return 1 + max(minDepth(root.Left), minDepth(root.Right))
}

File: minimum-depth-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
# depth is min depth of (left, right), when both exist
# otherwise, it is max depth of left, right; because a zero depth is not a depth
# 1 
#   2
#     3
class Solution:
    def minDepth(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        if root.left and root.right:
            return 1 + min(self.minDepth(root.left), self.minDepth(root.right))
        return 1 + max(self.minDepth(root.left), self.minDepth(root.right))
        
